### PR TITLE
Do not kill udev during boot

### LINKED ISSUE #39867: AWS KVM instances fail to boot because NVMe device doesn't appear
## Issue description
AWS EC2 KVM instances (M5, C5) may crash with kernel panic during boot.

### Steps to reproduce
launch nixos ami (tested on 17.09 and 18.03) and select m5.large instance type.
Set root volume to 200Gb (increases chances to reproduce)

## Technical details
Instance's system log may show from 1 up to many crash-loops. Eventually instance will boot, but it may take up to 30 minutes. It is not 100% reproducible scenario, so use bulk testing to get a repro:

I launched 5 identical instances m5.xlarge 200Gb root volume
One instance came up perfectly fine and fast. No kernel panics at all
Second one, also came up pretty fast. 1 Kernel panic, but second boot loop was fine
Third one - had 12 kernel panics, booted up after 10 minutes
Other 2 - were still not available ~20 minutes after launch

Amazon Linux AMIs behave normally which makes me believe it is an OS configuration error.

Example of system log showing kernel panic.
```
<<< NixOS Stage 1 >>>
loading module xen-blkfront...
modprobe: ERROR: could not insert 'xen_blkfront': No such device
loading module xen-netfront...
modprobe: ERROR: could not insert 'xen_netfront': No such device
loading module fuse...
loading module dm_mod...
loading module af_packet...
running udev...
kbd_mode: KDSKBMODE: Inappropriate ioctl for device
%Gbringing up network interface eth0...
bringing up network interface lo...
acquiring IP address via DHCP...
udhcpc: started, v1.28.3
udhcpc: sending discover
udhcpc: sending select for x.x.x.x
udhcpc: lease of x.x.x.x obtained, lease time 3600
networking is up!
starting device mapper and LVM...
waiting for device /dev/disk/by-label/nixos to appear.......................
Timed out waiting for device /dev/disk/by-label/nixos, trying to mount anyway.
mounting /dev/disk/by-label/nixos on /...
mount: mounting /dev/disk/by-label/nixos on /mnt-[ro ot /  fa2il2ed.: 1N62432] Kernel panic - not syncing: Attempted to kill init! exitcode=0x00000100
[   22.162432] 
[   22.162990] stage-1-init: mount: mounting /dev/disk/by-label/nixos on /mnt-root/ failed: No such file or directory
o such[ f il e  or2 d2.165776] CPU: 1 PID: 1 Comm: init Tainted: G           O    4.14.38 #1-NixOS
[   22.166913] Hardware name: Amazon EC2 m5.large/, BIOS 1.0 10/16/2017
irec[to ry 
22.167936] Call Trace:
[   22.168068] random: fast init done
[   22.169003]  dump_stack+0x5c/0x85
[   22.169572]  panic+0xd5/0x232
[   22.170119]  do_exit+0xad0/0xad0
[   22.170641]  ? handle_mm_fault+0xaa/0x1e0
[   22.171365]  do_group_exit+0x3a/0xa0
[   22.171926]  SyS_exit_group+0x10/0x10
[   22.172549]  do_syscall_64+0x67/0x120
[   22.173144]  entry_SYSCALL_64_after_hwframe+0x3d/0xa2
[   22.174025] RIP: 0033:0x7fa0bfe9c676
[   22.174593] RSP: 002b:00007fff6f440128 EFLAGS: 00000202 ORIG_RAX: 00000000000000e7
[   22.175856] RAX: ffffffffffffffda RBX: 00000000015707a0 RCX: 00007fa0bfe9c676
[   22.176931] RDX: 0000000000000001 RSI: 000000000000003c RDI: 0000000000000001
[   22.178116] RBP: 0000000000000004 R08: 00000000000000e7 R09: ffffffffffffff80
[   22.179151] R10: 0000000000000000 R11: 0000000000000202 R12: 00007fff6f4404d8
[   22.180342] R13: 00007fff6f4404d0 R14: 0000000000000000 R15: 0000000000000000
[   22.181461] Kernel Offset: 0x3e00000 from 0xffffffff81000000 (relocation range: 0xffffffff80000000-0xffffffffbfffffff)
[   22.182974] Rebooting in 1 seconds..
<<< NixOS Stage 1 >>>
....
... boot loop repeats and crashes again many times...
... until eventually it succeeds
....
<<< NixOS Stage 1 >>>
loading module xen-blkfront...
modprobe: ERROR: could not insert 'xen_blkfront': No such device
loading module xen-netfront...
modprobe: ERROR: could not insert 'xen_netfront': No such device
loading module fuse...
loading module dm_mod...
loading module af_packet...
running udev...
kbd_mode: KDSKBMODE: Inappropriate ioctl for device
%Gbringing up network interface eth0...
bringing up network interface lo...
acquiring IP address via DHCP...
udhcpc: started, v1.28.3
udhcpc: sending discover
udhcpc: sending select for x.x.x.x
udhcpc: lease of x.x.x.x obtained, lease time 3600
networking is up!
starting device mapper and LVM...
CHANGED: partition=1 start=2048 old: size=4190208 end=4192256 new: size=419428319,end=419430367
checking /dev/disk/by-label/nixos...
fsck (busybox 1.28.3)
[fsck.ext4 (1) -- /mnt-root/] fsck.ext4 -a /dev/disk/by-label/nixos
nixos: clean, 67983/131072 files, 264020/523776 blocks
resizing /dev/disk/by-label/nixos...
nixos: 67983/131072 files (0.1% non-contiguous), 264020/523776 blocks
resize2fs 1.44.1 (24-Mar-2018)
Resizing the filesystem on /dev/disk/by-label/nixos to 52428539 (4k) blocks.
The filesystem on /dev/disk/by-label/nixos is now 52428539 (4k) blocks long.
mounting /dev/disk/by-label/nixos on /...
...
```

My info (but also happens on 18.03, 17.09)
```
 - system: `"x86_64-linux"`
 - host os: `Linux 4.14.38, NixOS, 18.09.git.2370440 (Jellyfish)`
 - multi-user?: `no`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.0.1`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos/nixpkgs`
```
P.S.

Also, per [AWS Docs](https://docs.aws.amazon.com/AWSEC2/latest/UserGuide/nvme-ebs-volumes.html#timeout-nvme-ebs-volumes) I also tried updating `nvme_core.io_timeout` value and 
a newer kernel 
nixos/maintainers/scripts/ec2/amazon-image.nix:
```
config.boot.kernelPackages = pkgs.linuxPackages_4_15;
config.boot.kernelParams = [ "nvme_core.io_timeout=255" ];
```
Did not change anything at all.

### PR DESCRIPTION
**Pull request by ngortheone**
Fixes https://github.com/NixOS/nixpkgs/issues/39867

So main culprit was `udevadm control --exit || true` during initrd stage. I don't know what kind of fix was that, but it is definitely outdated, and in generally killing udev is not a good idea.
Basically what was happening - when nvme device did not appear immediately and udev was killed there were no links to device created in `/dev/device-by*` that was causing inability to mount root volume. 

NVME devices may appear with some delay, so I had to add `waitDevice` to growpartition script, to ensure that root volume is resized even when device is delayed.

Tested on m5 and m4 instances.

### CODE REVIEW

File: nixos/modules/system/boot/stage-1-init.sh [hunk 1]
```diff
@@ -74,6 +74,32 @@ ln -s /proc/mounts /etc/mtab # to shut up mke2fs
...
+            udevadm trigger --action=add
+            if test -e $device; then break; fi
+            echo -n "."
+            try=$((try - 1))
+        done
+        echo
+        [ $try -ne 0 ]
+    fi
+}
+
```
Comment (xeji) [inline nixos/modules/system/boot/stage-1-init.sh]: This file looks like you just moved  `waitDevice()` around without making any real change. What's the point?

**Comment by ngortheone** [comment]
@edolstra @dezgeg

File: nixos/modules/system/boot/stage-1-init.sh (hunk 1 above)
Comment (ngortheone) [reply nixos/modules/system/boot/stage-1-init.sh]: yes, i needed it to be declared before `@postDeviceCommands@` so I can use the func in growPartition

**Comment by copumpkin** [comment]
@dezgeg so is not killing udev likely to bite us in other subtle ways later? It definitely seemed to be hurting our NVMe discovery on EC2, but it's unclear to me how to make sure the other issue doesn't pop up again. We already ask it to settle at various points along the way.

**Comment by dezgeg** [comment]
Yeah, probably the `settle`s that are added are now enough.

**Comment by xeji** [comment]
Does this need a backport to 18.03 as well?

**Comment by edolstra** [comment]
Looks good to me.

I would hold off on backporting to 18.03 until we feel confident that the race with udev that prompted the addition of the `udevadm control --exit` is really gone.

**Comment by dezgeg** [comment]
I would think the udev blkid race could be avoided by wrapping certain parts in `udevadm control --stop-exec-queue` and `udevadm control --start-exec-queue` as was proposed in the scylladb thread (and in fact would be the only way to fix the similar races that sometimes happen in NixOS installer tests).

**Comment by dezgeg** [comment]
For reference (to confirm that I am not talking crazy stuff :), these are the occasional failures I see in the installer tests which I believe to be related to the udev race: https://nix-cache.s3.amazonaws.com/log/rwfcwv4p18sbbnd7pna2r2m9aqz7s2m5-vm-test-run-installer-luksroot.drv

````
machine: must succeed: parted --script /dev/vda -- mkpart primary linux-swap 50M 1024M
machine# [   35.973121] systemd[1]: Started Networking Setup.
machine# [   35.978131] systemd[1]: Starting Extra networking commands....
machine# [   36.549789]  vda: vda1 vda2
machine# Error: Partition(s) 2 on /dev/vda have been written, but we have been unable to inform the kernel of the change, probably because it/they are in use.  As a result, the old partition(s) will remain in use.  You should reboot now before making further changes.
````

**Pull request merged**